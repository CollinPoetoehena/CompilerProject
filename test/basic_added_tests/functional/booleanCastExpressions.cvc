extern void printInt(int val);
extern void printNewlines(int num);
extern void printFloat( float val);

void printBool(bool b) {
    if(b) {
        printInt(1);
    } else {
        printInt(0);
    }
}

export int main() {
    int a=1;
    int b=2;
    int c=3;
    int d=4;
    float w=1.0;
    float x=2.0;
    bool q = true;
    bool z = false;
    int falseInt = 0;
    int trueInt = 1;
    float falseFloat = 0.7; // Everything smaller than 1 is false
    float trueFloat = 1.0; // Everything higher than or equal 1 is true

    // Boolean cast results
    d = (int)(a<b&&c>d); //d = 0
    printInt(d);
    printNewlines(1);

    d = (int)(b<d); //d = 1
    printInt(d);
    printNewlines(1);

    w = (float)(b>d||q); // w = 1.0
    printFloat(w);
    printNewlines(1);

    w = (float)(b>d&&q); // w = 0.0
    printFloat(w);
    printNewlines(1);

    w = (float)true;
    printFloat(w);
    printNewlines(1);

    d = (int)false;
    printInt(d);
    printNewlines(1);

    // Nested bool casts
    d = (int) ((bool)1);
    printInt(d);
    printNewlines(1);

    d = (int) boolFun();
    printInt(d);
    printNewlines(1);

    d = (int) ((bool) ((int) true));
    printInt(d);
    printNewlines(1);

    z = (bool) ((int)(a<b&&c>d));
    printBool(z);
    printNewlines(1);

    // Boolean cast results
    q = (bool)falseInt; // q = false
    printBool(q);
    printNewlines(1);

    q = (bool)trueInt; // q = true
    printBool(q);
    printNewlines(1);

    z = (bool)falseFloat; // z = false
    printBool(z);
    printNewlines(1);

    z = (bool)trueFloat; // z = true
    printBool(z);
    printNewlines(1);

    // variables with a bool type
    falseInt = (int)q;
    printInt(falseInt);
    printNewlines(1);

    trueFloat = (float)z;
    printFloat(trueFloat);
    printNewlines(1);

    // These should not be converted (useless casts)
    d = (int) 5;
    printInt(d);
    printNewlines(1);

    w = (float) 5.3;
    printFloat(w);
    printNewlines(1);

    d = (int) intFun();
    printInt(d);

    return 0;
}

bool boolFun() {
    return true;
}

int intFun() {
    return 5;
}