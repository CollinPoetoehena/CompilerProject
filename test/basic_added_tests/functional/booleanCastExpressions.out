0
0
1.000000
1.000000
1.000000
0
1
1
1
1
0
1
0
1
1
1.000000
5
5.300000
5